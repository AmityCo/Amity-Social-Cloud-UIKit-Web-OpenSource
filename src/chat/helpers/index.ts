export { addChatMembers } from './addChatMembers';
export { removeChatMembers } from './removeChatMembers';
