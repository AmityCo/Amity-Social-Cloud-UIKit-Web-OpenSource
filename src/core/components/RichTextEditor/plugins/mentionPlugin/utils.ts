import { ELEMENT_MENTION } from '@udecode/plate';
import { MentionData, MentionItem, MentionType } from './models';
import { MentionElement, Text } from '../../models';
import { MentionSymbol, MentionRegex } from './constants';

export function toMentionItem(data: MentionData): MentionItem {
  return {
    data,
    key: data.id,
    text: data.display,
  };
}

export function fromMentionItem(item: MentionItem): MentionData {
  return item.data;
}

export function mentionElementToStringWithTags(element: MentionElement) {
  return `${MentionSymbol[element.mentionType]}[${element.value}](${element.id})`;
}

export function mentionElementToStringWithoutTags(element: MentionElement) {
  return `${MentionSymbol[element.mentionType]}${element.value}`;
}

export function stripMentionTags(text: string) {
  return text.replaceAll(MentionRegex, '$1$2');
}

/** Obfuscate mention tags so the serializer does not turn them to links */
export function obfuscateMentionTags(text: string) {
  return text.replaceAll(MentionRegex, '$1($2)($3)');
}

/** Break text nodes with mention tags into mention elements */
export function breakTextNodeIntoMentions(
  textNode: Text,
  mentionWithTagsRegex: RegExp,
): Text | (Text | MentionElement)[] {
  if (!mentionWithTagsRegex.test(textNode.text)) {
    return textNode;
  }

  const nodes: (Text | MentionElement)[] = [];
  let currentIndex = 0;

  const matches = [...textNode.text.matchAll(new RegExp(mentionWithTagsRegex, 'g'))];
  matches.forEach((match) => {
    const matchIndex = textNode.text.indexOf(match[0]);
    nodes.push({ text: textNode.text.substring(currentIndex, matchIndex) });
    nodes.push({
      children: [{ text: '' }],
      type: ELEMENT_MENTION,
      mentionType: Object.keys(MentionSymbol).find(
        (key) => MentionSymbol[key] === match[1],
      ) as MentionType,
      value: match[2],
      id: match[3],
    });
    currentIndex = matchIndex + match[0].length;
  });

  nodes.push({ text: textNode.text.substring(currentIndex) });

  return nodes;
}
