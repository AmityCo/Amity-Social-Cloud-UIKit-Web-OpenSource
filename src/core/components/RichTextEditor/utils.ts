/* eslint-disable no-param-reassign */
import { Transforms, Descendant, Editor as SlateEditor, BaseEditor } from 'slate';
import { ReactEditor } from 'slate-react';
import isEqual from 'lodash.isequal';

import {
  TElement,
  wrapNodes,
  insertNodes,
  isCollapsed,
  unwrapNodes,
  isEditor,
  isElement,
  ELEMENT_LINK,
  getPluginType,
  someNode,
  getEditorString,
  insertText as insertEditorText,
} from '@udecode/plate';
import { Editor } from './models';

import { EMPTY_VALUE } from './constants';

export function getSelectedText(editor: Editor) {
  if (editor.selection) {
    return getEditorString(editor, editor.selection);
  }

  return '';
}

export function isEmptyValue(value?: Descendant[]) {
  return value?.length === 1 && isEqual(EMPTY_VALUE, value[0]);
}

export function insertText(editor: Editor, text: string) {
  insertEditorText(editor, text);
}

export function isLinkActive(editor: Editor) {
  const type = getPluginType(editor, ELEMENT_LINK);
  const isLink = !!editor?.selection && someNode(editor, { match: { type } });

  return isLink;
}

export function removeLink(editor: Editor) {
  unwrapNodes(editor, {
    match: (n) => !isEditor(n) && isElement(n) && n.type === ELEMENT_LINK,
  });
}

export function insertLink(editor: Editor, url: string, text?: string) {
  if (!editor.selection) {
    return;
  }

  if (isLinkActive(editor)) {
    removeLink(editor);
  }

  const type = getPluginType(editor, ELEMENT_LINK);

  if (isCollapsed(editor.selection)) {
    insertNodes<TElement>(editor, {
      type,
      link: url,
      children: [{ text: text || url }],
    });

    return;
  }
  unwrapNodes(editor, { at: editor.selection, match: { type } });

  wrapNodes(
    editor,
    {
      type: type as any,
      link: url,
      children: [],
    },
    { at: editor.selection, split: true },
  );
}

export function resetSelection(editor: ReactEditor) {
  const point = { path: [0, 0], offset: 0 };
  Transforms.select(editor, { anchor: point, focus: point });
}

export function calculateRowStyles(rows?: number, maxRows?: number) {
  if (!rows && !maxRows) {
    return {};
  }

  const adjustedRows = rows && maxRows && rows > maxRows ? maxRows : rows;

  return {
    overflow: 'auto',
    minHeight: adjustedRows ? `${adjustedRows * 1.3}em` : undefined,
    maxHeight: maxRows ? `${maxRows * 1.3}em` : undefined,
  };
}

export function clear(editor: Editor) {
  Transforms.delete(editor as BaseEditor, {
    at: {
      anchor: SlateEditor.start(editor as BaseEditor, []),
      focus: SlateEditor.end(editor as BaseEditor, []),
    },
  });
}
