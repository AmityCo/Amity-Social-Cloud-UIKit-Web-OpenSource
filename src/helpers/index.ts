export { default as isEqual } from './isEqual';
export { default as isEmpty } from './isEmpty';
export { default as readFileAsync } from './readFileAsync';
export {
  default as getCssPosition,
  POSITION_BOTTOM,
  POSITION_LEFT,
  POSITION_RIGHT,
  POSITION_TOP,
} from './getCssPosition';
