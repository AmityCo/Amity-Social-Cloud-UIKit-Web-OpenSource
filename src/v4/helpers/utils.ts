import { CommunityPostSettings } from '@amityco/ts-sdk';
import isEmpty from 'lodash/isEmpty';

export type Mentioned = {
  userId?: string;
  length: number;
  index: number;
  type: string;
  displayName?: string;
};
export type Mentionees = (Amity.UserMention | Amity.ChannelMention)[];
export type Metadata = {
  mentioned?: Mentioned[];
};

type Chunk = { start: number; end: number; highlight?: boolean };

export const processChunks = (text: string, chunks: Chunk[] = []) => {
  const textLength = text?.length || 0;
  const allChunks: Chunk[] = [];

  const append = (start: number, end: number, highlight: boolean) => {
    allChunks.push({ start, end, highlight });
  };

  if (!chunks || chunks?.length === 0) {
    append(0, textLength, false);
  } else {
    let lastIndex = 0;
    chunks.forEach((chunk) => {
      append(lastIndex, chunk.start, false);
      append(chunk.start, chunk.end, true);
      lastIndex = chunk.end;
    });
    append(lastIndex, textLength, false);
  }

  return allChunks;
};

export function isCommunityMember(member?: Amity.Member<'community'>) {
  return member?.communityMembership === 'member';
}

export function isCommunityPost(post?: Amity.Post) {
  return post?.targetType === 'community';
}

export function isPostUnderReview(post?: Amity.Post, community?: Amity.Community | null) {
  if (community?.postSetting === CommunityPostSettings.ANYONE_CAN_POST) {
    return false;
  }
  return true;
}

export function isPrivateNetwork(socialSettings?: Amity.SocialSettings | null) {
  return (socialSettings?.userPrivacySetting as any) === 'private';
}

// Because community users have another type
export function formatCommunityMentionees(mentionees: Amity.User[], users: Amity.User[]) {
  return (mentionees || []).map((mentionee) => {
    const user = users.find(({ userId }) => userId === mentionee.userId);
    if (user == null) return null;
    return {
      id: user.userId,
      display: user.displayName,
      avatar: user.avatarCustomUrl,
    };
  });
}

export function formatMentionees(users: Amity.User[] | undefined) {
  return (users || []).map((user, index) => ({
    id: user.userId,
    display: user.displayName ?? user.userId,
    avatar: user.avatarCustomUrl,
    isLastItem: users?.length === index + 1,
  }));
}

export function searchWords(mentionees: Mentionees) {
  // Prolly need to dissect this on upper level TextContent

  if (mentionees?.length) {
    if (Object.prototype.hasOwnProperty.call(mentionees[0], 'userIds')) {
      return (mentionees[0] as Amity.UserMention).userIds.map((userId) => `@${userId}`);
    }

    return ['@all'];
  }

  return [];
}

const AT_SIGN_LENGTH = 1;

export function extractDisplayName(
  text: string,
  displayNameLength: number,
  startOfMention: number,
) {
  const startOfName = startOfMention + AT_SIGN_LENGTH;
  const endOfName = startOfName + displayNameLength;

  return text.substring(startOfName, endOfName);
}

export function findChunks(mentionees?: Mentioned[]) {
  if (!mentionees) return [];

  const mentioneeChunks = mentionees
    .sort((m1, m2) => m1.index - m2.index)
    .map(({ index, length }) => ({
      start: index,
      end: index + length + AT_SIGN_LENGTH,
      highlight: true,
    }));

  return mentioneeChunks;
}

export function extractMetadata(
  mentions: { plainTextIndex: number; id: string; display: string }[],
) {
  const metadata: Metadata = {};
  const mentionees: (Amity.UserMention | Amity.ChannelMention)[] = [];

  if (mentions?.length > 0) {
    metadata.mentioned = [
      ...mentions
        .sort((m1, m2) => m1.plainTextIndex - m2.plainTextIndex)
        .map(({ plainTextIndex, id, display: displayName }) => {
          if (id === '@all') {
            return {
              index: plainTextIndex,
              length: displayName.length - AT_SIGN_LENGTH,
              type: 'channel',
            };
          }

          return {
            index: plainTextIndex,
            length: displayName.length - AT_SIGN_LENGTH,
            type: 'user',
            userId: id,
          };
        }),
    ];

    const isContainMentionAll = mentions.some(({ id }) => id === '@all');

    if (isContainMentionAll) {
      mentionees.push({ type: 'channel' } as Amity.ChannelMention);
    }

    mentionees.push({
      type: 'user',
      userIds: mentions.filter(({ id }) => id !== '@all').map(({ id }) => id),
    } as Amity.UserMention);
  } else {
    // to clean the mentionees on the backend side
    mentionees.push({ type: 'user', userIds: [] } as Amity.UserMention);
  }

  return { metadata, mentionees };
}

export function parseMentionsMarkup(
  text?: string,
  metadata?: { mentioned?: Array<{ userId: string; length: number; index: number; type: string }> },
) {
  if (!text) return text;

  if (isEmpty(metadata?.mentioned)) {
    return text;
  }

  let parsedText = text;

  ((metadata || {}).mentioned || [])
    .sort((a, b) => b.index - a.index)
    .forEach(({ userId, length, index: textIndex }) => {
      const markupFormat = `@[${extractDisplayName(text, length, textIndex)}](${userId})`;

      parsedText = parsedText.replace(
        new RegExp(`(.{${textIndex}}).{${length + AT_SIGN_LENGTH}}`),
        `$1${markupFormat}`,
      );
    });

  return parsedText;
}

export function isNonNullable<TValue>(value: TValue | undefined | null): value is TValue {
  return value != null;
}

export function getCssVariableValue(variable: string) {
  return getComputedStyle(document.documentElement).getPropertyValue(variable);
}

export function convertRemToPx(rem: number) {
  return rem * parseFloat(getComputedStyle(document.documentElement).fontSize);
}
/**
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 */

const SUPPORTED_URL_PROTOCOLS = new Set(['http:', 'https:', 'mailto:', 'sms:', 'tel:']);

export function sanitizeUrl(url: string): string {
  try {
    const parsedUrl = new URL(url);

    if (!SUPPORTED_URL_PROTOCOLS.has(parsedUrl.protocol)) {
      return 'about:blank';
    }
  } catch {
    return url;
  }
  return url;
}

// Source: https://stackoverflow.com/a/8234912/2013580
const urlRegExp = new RegExp(
  /((([A-Za-z]{3,9}:(?:\/\/)?)(?:[-;:&=+$,\w]+@)?[A-Za-z0-9.-]+|(?:www.|[-;:&=+$,\w]+@)[A-Za-z0-9.-]+)((?:\/[+~%/.\w-_]*)?\??(?:[-+=&;%@.\w_]*)#?(?:[\w]*))?)/,
);
export function validateUrl(url: string): boolean {
  // TODO Fix UI for link insertion; it should never default to an invalid URL such as https://.
  // Maybe show a dialog where they user can type the URL before inserting it.
  return url === 'https://' || urlRegExp.test(url);
}
