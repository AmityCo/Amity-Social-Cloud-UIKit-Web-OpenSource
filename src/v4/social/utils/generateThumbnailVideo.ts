export const generateThumbnailVideo = (file: File): Promise<string> => {
  return new Promise((resolve, reject) => {
    const videoElement = document.createElement('video');
    const canvasElement = document.createElement('canvas');
    const context = canvasElement.getContext('2d');

    videoElement.src = URL.createObjectURL(file);

    const handleCanPlay = () => {
      videoElement.currentTime = 10;
    };

    const handleSeeked = () => {
      try {
        canvasElement.width = videoElement.videoWidth;
        canvasElement.height = videoElement.videoHeight;
        context?.drawImage(videoElement, 0, 0, canvasElement.width, canvasElement.height);
        const thumbnail = canvasElement.toDataURL('image/png');
        resolve(thumbnail);
      } catch (error) {
        reject(error);
      } finally {
        videoElement.removeEventListener('canplay', handleCanPlay);
        videoElement.removeEventListener('seeked', handleSeeked);
        videoElement.removeEventListener('error', handleError);
        URL.revokeObjectURL(videoElement.src);
      }
    };

    const handleError = (event: Event) => {
      reject(new Error(`Error loading video: ${event}`));
    };

    videoElement.addEventListener('canplay', handleCanPlay);
    videoElement.addEventListener('seeked', handleSeeked);
    videoElement.addEventListener('error', handleError);

    videoElement.load();
  });
};
