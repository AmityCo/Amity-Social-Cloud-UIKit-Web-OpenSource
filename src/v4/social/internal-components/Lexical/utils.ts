import { SerializedAutoLinkNode } from '@lexical/link';
import { InitialConfigType, InitialEditorStateType } from '@lexical/react/LexicalComposer';
import {
  EditorThemeClasses,
  Klass,
  LexicalEditor,
  LexicalNode,
  SerializedLexicalNode,
  SerializedParagraphNode,
  SerializedRootNode,
  SerializedTextNode,
} from 'lexical';
import { Mentioned, Mentionees } from '~/v4/helpers/utils';
import { SerializedMentionNode } from './nodes/MentionNode';

export interface EditorStateJson extends SerializedLexicalNode {
  children: [];
}

export function $isSerializedTextNode(node: SerializedLexicalNode): node is SerializedTextNode {
  return node.type === 'text';
}

export function $isSerializedMentionNode<T>(
  node: SerializedLexicalNode,
): node is SerializedMentionNode<T> {
  return node.type === 'mention';
}

export function $isSerializedAutoLinkNode(
  node: SerializedLexicalNode,
): node is SerializedAutoLinkNode {
  return node.type === 'autolink';
}

export type MentionData = {
  userId: string;
  displayName?: string;
};

function createRootNode(): SerializedRootNode<SerializedParagraphNode> {
  return {
    children: [],
    direction: 'ltr',
    format: '',
    indent: 0,
    type: 'root',
    version: 1,
  };
}

function createParagraphNode(): SerializedParagraphNode {
  return {
    children: [],
    direction: 'ltr',
    format: '',
    indent: 0,
    type: 'paragraph',
    version: 1,
    textFormat: 0,
  };
}

function createSerializeTextNode(text: string): SerializedTextNode {
  return {
    detail: 0,
    format: 0,
    mode: 'normal',
    style: '',
    text,
    type: 'text',
    version: 1,
  };
}

function createSerializeMentionNode(mention: Mentioned): SerializedMentionNode<MentionData> {
  return {
    detail: 0,
    format: 0,
    mode: 'normal',
    style: '',
    text: ('@' + mention.userId) as string,
    type: 'mention',
    version: 1,
    data: {
      displayName: mention.userId as string,
      userId: mention.userId as string,
    },
  };
}

export function textToEditorState(value: {
  data: { text: string };
  metadata?: {
    mentioned?: Mentioned[];
  };
  mentionees?: Mentionees;
}) {
  const rootNode = createRootNode();

  const textArray = value.data.text.split('\n');

  const mentions = value.metadata?.mentioned || [];

  let mentionIndex = 0;
  let globalIndex = 0;

  for (let i = 0; i < textArray.length; i += 1) {
    const paragraph = createParagraphNode();
    let runningIndex = 0;

    while (runningIndex < textArray[i].length) {
      if (mentionIndex < mentions.length && mentions[mentionIndex].index === globalIndex) {
        paragraph.children.push(createSerializeMentionNode(mentions[mentionIndex]));
        runningIndex += mentions[mentionIndex].length;
        globalIndex += mentions[mentionIndex].length;
        mentionIndex += 1;
      } else {
        const nextMentionIndex =
          mentionIndex < mentions.length
            ? mentions[mentionIndex].index
            : globalIndex + textArray[i].length;
        const textSegment = textArray[i].slice(
          runningIndex,
          nextMentionIndex - globalIndex + runningIndex,
        );
        if (textSegment) {
          paragraph.children.push(createSerializeTextNode(textSegment));
        }
        runningIndex += textSegment.length;
        globalIndex += textSegment.length;
      }
    }

    if (runningIndex < textArray[i].length) {
      const textSegment = textArray[i].slice(runningIndex);
      if (textSegment) {
        paragraph.children.push(createSerializeTextNode(textSegment));
      }
      globalIndex += textSegment.length;
    }

    globalIndex += 1;

    rootNode.children.push(paragraph);
  }

  return { root: rootNode };
}

export function editorStateToText(editor: LexicalEditor) {
  const editorStateTextString: string[] = [];
  const paragraphs = editor.getEditorState().toJSON().root.children as EditorStateJson[];

  const mentioned: Mentioned[] = [];
  let isChannelMentioned = false;
  const mentioneeUserIds: string[] = [];
  let runningIndex = 0;

  paragraphs.forEach((paragraph) => {
    const children = paragraph.children;
    const paragraphText: string[] = [];

    children.forEach(
      (child: SerializedTextNode | SerializedMentionNode<MentionData> | SerializedAutoLinkNode) => {
        if ($isSerializedTextNode(child)) {
          paragraphText.push(child.text);
          runningIndex += child.text.length;
        }
        if ($isSerializedAutoLinkNode(child)) {
          child.children.forEach((c) => {
            if (!$isSerializedTextNode(c)) return;
            paragraphText.push(c.text);
            runningIndex += c.text.length;
          });
        }

        if ($isSerializedMentionNode<MentionData>(child)) {
          if (child.data.userId === 'all') {
            mentioned.push({
              index: runningIndex,
              length: child.text.length,
              type: 'channel',
            });
            isChannelMentioned = true;
          } else {
            mentioned.push({
              index: runningIndex,
              length: child.text.length,
              type: 'user',
              userId: child.data.userId,
            });
            mentioneeUserIds.push(child.data.userId);
          }
          paragraphText.push(child.text);
          runningIndex += child.text.length;
        }
      },
    );
    runningIndex += 1;
    editorStateTextString.push(paragraphText.join(''));
  });

  const mentionees: Array<Amity.UserMention | Amity.ChannelMention> = [];

  if (mentioneeUserIds.length > 0) {
    mentionees.push({
      type: 'user',
      userIds: mentioneeUserIds,
    });
  }

  if (isChannelMentioned) {
    mentionees.push({ type: 'channel' });
  }

  return { mentioned, text: editorStateTextString.join('\n'), mentionees };
}

const defaultTheme = {
  ltr: 'ltr',
  rtl: 'rtl',
};

export const getEditorConfig = ({
  namespace,
  theme,
  nodes,
  editorState,
}: {
  namespace: string;
  theme: EditorThemeClasses;
  nodes: Array<Klass<LexicalNode>>;
  editorState?: InitialEditorStateType;
}): InitialConfigType => ({
  namespace,
  editable: true,
  theme: { ...defaultTheme, ...theme },
  onError(error: Error) {
    throw error;
  },
  nodes,
  editorState,
});
