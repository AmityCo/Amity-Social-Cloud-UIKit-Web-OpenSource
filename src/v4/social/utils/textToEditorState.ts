import { SerializedParagraphNode, SerializedRootNode, SerializedTextNode } from 'lexical';
import { Mentioned, Mentionees } from '~/v4/helpers/utils';
import { SerializedMentionNode } from '~/v4/social/internal-components/MentionTextInput/MentionNodes';

function createRootNode(): SerializedRootNode<SerializedParagraphNode> {
  return {
    children: [],
    direction: 'ltr',
    format: '',
    indent: 0,
    type: 'root',
    version: 1,
  };
}
function createParagraphNode(): SerializedParagraphNode {
  return {
    children: [],
    direction: 'ltr',
    format: '',
    indent: 0,
    type: 'paragraph',
    version: 1,
    textFormat: 0,
  };
}

function createSerializeTextNode(text: string): SerializedTextNode {
  return {
    detail: 0,
    format: 0,
    mode: 'normal',
    style: '',
    text,
    type: 'text',
    version: 1,
  };
}

function createSerializeMentionNode(mention: Mentioned): SerializedMentionNode {
  return {
    detail: 0,
    format: 0,
    mode: 'normal',
    style: '',
    text: ('@' + mention.userId) as string,
    type: 'mention',
    version: 1,
    mentionName: mention.userId as string,
    displayName: mention.userId as string,
    userId: mention.userId as string,
    userInternalId: mention.userId as string,
    userPublicId: mention.userId as string,
  };
}

export function textToEditorState(value: {
  data: { text: string };
  metadata?: {
    mentioned?: Mentioned[];
  };
  mentionees?: Mentionees;
}) {
  const rootNode = createRootNode();
  const textArray = value.data.text.split('\n');
  const mentions = value.metadata?.mentioned || [];
  let mentionIndex = 0;
  let globalIndex = 0;

  for (let i = 0; i < textArray.length; i++) {
    const paragraph = createParagraphNode();
    let runningIndex = 0;

    while (runningIndex < textArray[i].length) {
      if (mentionIndex < mentions.length && mentions[mentionIndex].index === globalIndex) {
        paragraph.children.push(createSerializeMentionNode(mentions[mentionIndex]));
        runningIndex += mentions[mentionIndex].length;
        globalIndex += mentions[mentionIndex].length;
        mentionIndex++;
      } else {
        const nextMentionIndex =
          mentionIndex < mentions.length
            ? mentions[mentionIndex].index
            : globalIndex + textArray[i].length;
        const textSegment = textArray[i].slice(
          runningIndex,
          nextMentionIndex - globalIndex + runningIndex,
        );
        if (textSegment) {
          paragraph.children.push(createSerializeTextNode(textSegment));
        }
        runningIndex += textSegment.length;
        globalIndex += textSegment.length;
      }
    }

    if (runningIndex < textArray[i].length) {
      const textSegment = textArray[i].slice(runningIndex);
      if (textSegment) {
        paragraph.children.push(createSerializeTextNode(textSegment));
      }
      globalIndex += textSegment.length;
    }

    globalIndex += 1;

    rootNode.children.push(paragraph);
  }

  return { root: rootNode };
}
