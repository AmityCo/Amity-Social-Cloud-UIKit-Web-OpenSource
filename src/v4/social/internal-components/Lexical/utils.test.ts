import { Metadata, Mentionees } from '~/v4/helpers/utils';
import { editorStateToText, textToEditorState } from './utils';

type Input = {
  data: {
    text: string;
  };
  metadata: Metadata;
  mentionees: Mentionees;
};

const inputsV3: Input[] = [
  {
    data: {
      text: '@Web-Test hello @FonTS ',
    },
    metadata: {
      mentioned: [
        {
          index: 0,
          length: 8,
          type: 'user',
          userId: 'Web-Test',
        },
        {
          index: 16,
          length: 5,
          type: 'user',
          userId: 'FonTS',
        },
      ],
    },
    mentionees: [
      {
        type: 'user',
        userIds: ['Web-Test', 'FonTS'],
      },
    ],
  },
];

const inputsV4: Input[] = [
  {
    data: {
      text: '1 @ android2 nstaehunstah @ android2 @Web-Test aenotuhnasetouhnsateohunsatheunst sntaehuns\nSTnseatu @ android3 ',
    },
    metadata: {
      mentioned: [
        { index: 2, length: 9, type: 'user', userId: ' android2', displayName: ' android2' },
        { index: 26, length: 9, type: 'user', userId: ' android2', displayName: ' android2' },
        { index: 37, length: 8, type: 'user', userId: 'Web-Test', displayName: 'Web-Test' },
        { index: 100, length: 9, type: 'user', userId: ' android3', displayName: ' android3' },
      ],
    },
    mentionees: [{ type: 'user', userIds: [' android2', ' android2', 'Web-Test', ' android3'] }],
  },
  {
    data: {
      text: 'www.google.com\nwww.youtube.com',
    },
    metadata: {
      mentioned: [],
    },
    mentionees: [],
  },
];

describe('v3', () => {
  test.each(inputsV3)('should return a same object', (input) => {
    const editorState = textToEditorState(input);

    expect(editorState).toMatchSnapshot();

    const actual = editorStateToText(editorState);
    expect(actual.text).toEqual(input.data.text);
    const expectedMentioned = (input.metadata.mentioned || []).map((m) => ({
      ...m,
      displayName: m.userId,
    }));
    expect(actual.mentioned).toEqual(expectedMentioned);
    expect(actual.mentionees).toEqual(input.mentionees);
  });
});

describe('v4', () => {
  test.each(inputsV4)('should return a same object', (input) => {
    const editorState = textToEditorState(input);

    expect(editorState).toMatchSnapshot();

    const actual = editorStateToText(editorState);
    expect(actual.text).toEqual(input.data.text);
    const expectedMentioned = actual.mentioned.map(({ ...m }) => m);
    expect(expectedMentioned).toEqual(input.metadata.mentioned);
    expect(actual.mentionees).toEqual(input.mentionees);
  });
});
