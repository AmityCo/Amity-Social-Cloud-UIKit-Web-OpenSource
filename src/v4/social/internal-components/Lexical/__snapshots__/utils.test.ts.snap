// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`v3 should return a same object 1`] = `
{
  "root": {
    "children": [
      {
        "children": [
          {
            "data": {
              "displayName": "Web-Test",
              "userId": "Web-Test",
            },
            "detail": 0,
            "format": 0,
            "mode": "normal",
            "style": "",
            "text": "@Web-Test",
            "type": "mention",
            "version": 1,
          },
          {
            "detail": 0,
            "format": 0,
            "mode": "normal",
            "style": "",
            "text": " hello ",
            "type": "text",
            "version": 1,
          },
          {
            "data": {
              "displayName": "FonTS",
              "userId": "FonTS",
            },
            "detail": 0,
            "format": 0,
            "mode": "normal",
            "style": "",
            "text": "@FonTS",
            "type": "mention",
            "version": 1,
          },
          {
            "detail": 0,
            "format": 0,
            "mode": "normal",
            "style": "",
            "text": " ",
            "type": "text",
            "version": 1,
          },
        ],
        "direction": "ltr",
        "format": "",
        "indent": 0,
        "textFormat": 0,
        "textStyle": "",
        "type": "paragraph",
        "version": 1,
      },
    ],
    "direction": "ltr",
    "format": "",
    "indent": 0,
    "type": "root",
    "version": 1,
  },
}
`;

exports[`v4 should return a same object 1`] = `
{
  "root": {
    "children": [
      {
        "children": [
          {
            "detail": 0,
            "format": 0,
            "mode": "normal",
            "style": "",
            "text": "1 ",
            "type": "text",
            "version": 1,
          },
          {
            "data": {
              "displayName": " android2",
              "userId": " android2",
            },
            "detail": 0,
            "format": 0,
            "mode": "normal",
            "style": "",
            "text": "@ android2",
            "type": "mention",
            "version": 1,
          },
          {
            "detail": 0,
            "format": 0,
            "mode": "normal",
            "style": "",
            "text": " nstaehunstah ",
            "type": "text",
            "version": 1,
          },
          {
            "data": {
              "displayName": " android2",
              "userId": " android2",
            },
            "detail": 0,
            "format": 0,
            "mode": "normal",
            "style": "",
            "text": "@ android2",
            "type": "mention",
            "version": 1,
          },
          {
            "detail": 0,
            "format": 0,
            "mode": "normal",
            "style": "",
            "text": " ",
            "type": "text",
            "version": 1,
          },
          {
            "data": {
              "displayName": "Web-Test",
              "userId": "Web-Test",
            },
            "detail": 0,
            "format": 0,
            "mode": "normal",
            "style": "",
            "text": "@Web-Test",
            "type": "mention",
            "version": 1,
          },
          {
            "detail": 0,
            "format": 0,
            "mode": "normal",
            "style": "",
            "text": " aenotuhnasetouhnsateohunsatheunst sntaehuns",
            "type": "text",
            "version": 1,
          },
        ],
        "direction": "ltr",
        "format": "",
        "indent": 0,
        "textFormat": 0,
        "textStyle": "",
        "type": "paragraph",
        "version": 1,
      },
      {
        "children": [
          {
            "detail": 0,
            "format": 0,
            "mode": "normal",
            "style": "",
            "text": "STnseatu ",
            "type": "text",
            "version": 1,
          },
          {
            "data": {
              "displayName": " android3",
              "userId": " android3",
            },
            "detail": 0,
            "format": 0,
            "mode": "normal",
            "style": "",
            "text": "@ android3",
            "type": "mention",
            "version": 1,
          },
          {
            "detail": 0,
            "format": 0,
            "mode": "normal",
            "style": "",
            "text": " ",
            "type": "text",
            "version": 1,
          },
        ],
        "direction": "ltr",
        "format": "",
        "indent": 0,
        "textFormat": 0,
        "textStyle": "",
        "type": "paragraph",
        "version": 1,
      },
    ],
    "direction": "ltr",
    "format": "",
    "indent": 0,
    "type": "root",
    "version": 1,
  },
}
`;

exports[`v4 should return a same object 2`] = `
{
  "root": {
    "children": [
      {
        "children": [
          {
            "children": [
              {
                "detail": 0,
                "format": 0,
                "mode": "normal",
                "style": "",
                "text": "www.google.com",
                "type": "text",
                "version": 1,
              },
            ],
            "direction": "ltr",
            "format": "",
            "indent": 0,
            "rel": null,
            "target": null,
            "title": null,
            "type": "link",
            "url": "http://www.google.com",
            "version": 1,
          },
        ],
        "direction": "ltr",
        "format": "",
        "indent": 0,
        "textFormat": 0,
        "textStyle": "",
        "type": "paragraph",
        "version": 1,
      },
      {
        "children": [
          {
            "children": [
              {
                "detail": 0,
                "format": 0,
                "mode": "normal",
                "style": "",
                "text": "www.youtube.com",
                "type": "text",
                "version": 1,
              },
            ],
            "direction": "ltr",
            "format": "",
            "indent": 0,
            "rel": null,
            "target": null,
            "title": null,
            "type": "link",
            "url": "http://www.youtube.com",
            "version": 1,
          },
        ],
        "direction": "ltr",
        "format": "",
        "indent": 0,
        "textFormat": 0,
        "textStyle": "",
        "type": "paragraph",
        "version": 1,
      },
    ],
    "direction": "ltr",
    "format": "",
    "indent": 0,
    "type": "root",
    "version": 1,
  },
}
`;
